doc/crypto/overview/functionality.rst

```diff
@@ -256,32 +205,105 @@ The typical sequence of actions with a multi-part operation is as follows:
-object will enter the *error* state.
+    -   Set it to logical zero. This is automatic for static and global variables. Explicit initialization must use the associated ``PSA_xxx_IOP_INIT`` macro as the type is implementation-defined.
+    -   Set it to all-bits zero. This is automatic if the object was allocated with ``calloc()``.
+    -   Assign the value of the associated macro ``PSA_xxx_IOP_INIT``.
+    -   Assign the result of calling the associated function ``psa_xxx_iop_init()``.
 
-It is safe to move a multi-part operation object to a different memory location,
-for example, using a bitwise copy, and then to use the object in the new
-location. For example, an application can allocate an operation object on the
-stack and return it, or the operation object can be allocated within memory
-managed by a garbage collector. However, this does not permit the following
-behaviors:
+    The resulting object is now *inactive*.
+    It is an error to initialize an operation object that is in *active* or *error* states. This can leak memory or other resources.
+
+#.  **Begin-setup:** Start a new interruptible operation on an *inactive* operation object. Each operation object will define one or more setup functions to start a specific operation.
+
+    On success, an operation object enters a *setup* state. On failure, the operation object will remain *inactive*.
+
+#.  **Complete-setup:** Complete the operation setup on an interruptible operation object that is in *setup* state.
+
+    If the setup computation is interrupted, a the operation remains in *setup* state. If setup completes successfully, the operation enters an *input* state. On failure, the operation object will enter an *error* state.
+
+    An application needs to repeat this step until the setup completes with success or an error status.
+
+#.  **Update:** Update an interruptible operation object in *input* state. The update function can provide additional parameters, supply data for processing or generate outputs.
+
+    On success, the operation object remains in *input* state. On failure, the operation object will enter an *error* state.
+
+#.  **Complete:** To end an interruptible operation, call the applicable completion function. This will perform the final computation, produce any final outputs, and then release any resources associated with the operation.
+
+    If the finishing computation is interrupted, a the operation is left in *completing* state. If the operation completes successfully, the operation enters an *inactive* state. On failure, the operation object will enter an *error* state.
+
+    An application needs to repeat this step until the completion function completes with success or an error status.
+
+#.  **Abort:** An interruptible operation can be aborted at any stage during its use by calling the associated ``psa_xxx_iop_abort()`` function. This will release any resources associated with the operation and return the operation object to the *inactive* state.
```

Comment:
Also, abort resets the op count to 0, whereas complete doesn't.

This is stated in the description of the `get_num_ops()` functions, but not apparent from the text here. It's also not visible in the state diagram, but I'm not sure it should be: `get_num_ops()` is perfectly well-defined in the active state, the subtlety is that it doesn't return 0 if the inactive state was reached after a complete() rather than initialization or abort.

Reply:
I've addressed this to the updated Functionality chapter in PR #199, which extends this one to cover key agreement use cases.